### ISSUE TITLE
MinGW support

### ISSUE THREAD

**Issue by nxrighthere** [sctplab/usrsctp#276]
I see a bunch of pull requests related to MinGW support, but they are quite outdated. Any plans to support MinGW out of the box, please, to avoid headaches with digging the code?

**Comment by tuexen** [comment]
I picked from the patches what I found generic improvements. I found contradicting change requests in different pull request and, since we are not windows experts, have no MinWG environment. If I remember it correctly, I think when we were looking around to set one up, we found that there are multiple ways of doing so and it wasn't clear which one to use.

Is there one MinGW environment we should use? Getting something in I can't test is something I'm hesitating to do.

**Comment by CoolmanCZ** [comment]
Hi,
I think that [Appveyor](https://www.appveyor.com) can be used to test windows version (at least for the building). It has an integration with GitHub.

**Comment by nxrighthere** [comment]
> Is there one MinGW environment we should use? 

Most people are using [this](https://sourceforge.net/projects/mingw-w64/) distribution. 

It would be great if you can separate the code in the traditional way using directives like `#if defined(_WIN32) && defined(__GNUC__)` instead of `__MINGW32__` for better compatibility. Right now GCC throws a lot of errors. Some of them are quite easy to fix, but some are not, unfortunately.

> I think that Appveyor can be used to test windows version (at least for the building).

Thanks for the suggestion.

**Comment by CoolmanCZ** [comment]
If you are looking for mingw version of the usrsctp take a look at [my MinGW fork](https://github.com/CoolmanCZ/usrsctp). I've been trying to keep the source code updated with latest version of the usrsctp.

**Comment by nxrighthere** [comment]
@CoolmanCZ Thanks! I've built debug version without any problems, but release throws some errors, unfortunately.

**Comment by CoolmanCZ** [comment]
Hi, can you send me the command to build the release version?

**Comment by nxrighthere** [comment]
You can set it right in the CMake:

![release](https://user-images.githubusercontent.com/8279330/48259771-e372fc80-e43a-11e8-9ab6-6fbf68131c5f.PNG)

**Comment by CoolmanCZ** [comment]
OK, try fixed Mingw branch.

**Comment by nxrighthere** [comment]
That was fast, thank you, it works. 😺

**Comment by weinrank** [comment]
We already have a Windows buildbot with MinGW support, but we haven't used it yet.

I'll have a closer look at your branch during the next days and we'll check how to integrate it. 👍

**Comment by nxrighthere** [comment]
@weinrank Any progress on this? 🎗

**Comment by tuexen** [comment]
This should be fixed now. If not, please re-open.

**Comment by nxrighthere** [comment]
Thanks, just built the library. 👍 

There are only some problems with example programs like:
```
E:\C\SCTP\programs\programs_helper.c: In function 'handle_peer_address_change_event':
E:\C\SCTP\programs\programs_helper.c:134:10: error: implicit declaration of function 'inet_ntop'; did you mean 'inet_ntoa'? [-Werror=implicit-function-declaration]
   addr = inet_ntop(AF_INET, &sin->sin_addr, addr_buf, INET_ADDRSTRLEN);
          ^~~~~~~~~
          inet_ntoa
E:\C\SCTP\programs\programs_helper.c:134:8: error: assignment to 'const char *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
   addr = inet_ntop(AF_INET, &sin->sin_addr, addr_buf, INET_ADDRSTRLEN);
        ^
E:\C\SCTP\programs\programs_helper.c:138:8: error: assignment to 'const char *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
   addr = inet_ntop(AF_INET6, &sin6->sin6_addr, addr_buf, INET6_ADDRSTRLEN);
        ^
E:\C\SCTP\programs\programs_helper.c: In function 'handle_send_failed_event':
E:\C\SCTP\programs\programs_helper.c:200:31: error: format '%u' expects argument of type 'unsigned int', but argument 2 has type 'u_long' {aka 'long unsigned int'} [-Werror=format=]
  printf("message with PPID = %u, SID = %u, flags: 0x%04x due to error = 0x%08x",
                              ~^
                              %lu
         ntohl(ssfe->ssfe_info.snd_ppid), ssfe->ssfe_info.snd_sid,
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
E:\C\SCTP\programs\programs_helper.c:200:31: error: format '%u' expects argument of type 'unsigned int', but argument 2 has type 'u_long' {aka 'long unsigned int'} [-Werror=format=]
  printf("message with PPID = %u, SID = %u, flags: 0x%04x due to error = 0x%08x",
                              ~^
                              %lu
         ntohl(ssfe->ssfe_info.snd_ppid), ssfe->ssfe_info.snd_sid,
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1.exe: all warnings being treated as errors
make[2]: *** [programs\CMakeFiles\chargen_server_upcall.dir\build.make:77: programs/CMakeFiles/chargen_server_upcall.dir/programs_helper.c.obj] Error 1
make[1]: *** [CMakeFiles\Makefile2:226: programs/CMakeFiles/chargen_server_upcall.dir/all] Error 2
make: *** [Makefile:129: all] Error 2
```

**Comment by tuexen** [comment]
The `inet_ntop()` problem is related to a missing declaration, I guess. @weinrank : Can you take a look?

The other problem is a bug in gcc on Windows, I guess. `ntohl()` on windows returns an `unsigned long`, whereas `%u` expects an `unsigned int`. However, on Windows, both a 32-bit entities. So gcc should not issue a warning, but unfortunately does. That is why @weinrank disabled the corresponding warning in https://github.com/sctplab/usrsctp/pull/321/commits/111e0caa5911833ef223df0ec92546582c8d83c7. Are you saying that the cmake based build does not work on your system?

**Comment by nxrighthere** [comment]
> Are you saying that the cmake based build does not work on your system?

Yup, I'm using CMake for building.

**Comment by nxrighthere** [comment]
For `inet_ntop()` and `inet_pton()` you can grab https://github.com/CoolmanCZ/usrsctp/tree/mingw/programs/mingw from @CoolmanCZ fork. It's under ISC license.

**Comment by weinrank** [comment]
@nxrighthere Which MinGW environment do you use?
MSYS? OS?

**Comment by nxrighthere** [comment]
I'm using [MinGW-W64 GCC-8.1.0 x86_64-posix-seh](https://sourceforge.net/projects/mingw-w64/files/) on Windows directly with CMake.

**Comment by nxrighthere** [comment]
Can't build the library from master with the latest changes, too many errors.

**Comment by lgrahl** [comment]
Yeah. I really do suggest [taking my CI files](https://github.com/rawrtc/usrsctp/commit/f04f4eae42bfa591fc1e60866d5876c890f3397d), so you're able to detect breaking changes with each commit made. I can help you setting it up, @tuexen & @weinrank.

**Comment by tuexen** [comment]
@weinrank: Do we have MinGW on a buildbot? If yes, are we detecting the problems?
@nxrighthere What errors do you see?

**Comment by lgrahl** [comment]
I mean, being able to see the build status on GitHub, like here (click the X next to the commit): https://github.com/rawrtc/usrsctp/commits/meson-test

You can see the build error here: https://circleci.com/gh/rawrtc/usrsctp/191?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by weinrank** [comment]
@tuexen I have an MSYS/MinGW buildbot for internal use but it's not yet exposed to the public buildbot interface.
This is because I'm still not sure if I'm using the "correct" MinGW setup. I'm not an expert in this area.

I've modified and successfully tested the usrsctp library in a MSYS2 environment using the MinGW64 shell. That's what have been merged recently.

Now I've set up a fresh Windows host and installed `MinGW-W64 GCC-8.1.0 x86_64-posix-seh` @nxrighthere mentioned. I can see compiling issues regarding `inet_ntop`.
So there seems to be a significant difference between the MinGW Software itself and the MSYS2 distribution.

I'm currently looking into the problem.

@lgrahl 
Adding a MinGW buildbot to our existing setup isn't such a big deal, thanks for the offer!

**Comment by nxrighthere** [comment]
> @nxrighthere What errors do you see?

There's only one left now:
```
E:\C\SCTP\usrsctplib\netinet\sctp_pcb.c: In function 'sctp_pcb_init':
E:\C\SCTP\usrsctplib\netinet\sctp_pcb.c:6854:2: error: implicit declaration of function 'atomic_init'; did you mean 'atomic_add_int'? [-Werror=implicit-function-declaration]
  atomic_init();
  ^~~~~~~~~~~
  atomic_add_int
```

> I can see compiling issues regarding inet_ntop.

You can grab the implementation from [here](https://github.com/nxrighthere/ENet-CSharp/tree/master/Source/Native/mingw).

**Comment by tuexen** [comment]
@nxrighthere This should be fixed by https://github.com/sctplab/usrsctp/commit/1a0619db218f6804bfd3c420b71a4422a7da17d7. Does it work for you?

**Comment by nxrighthere** [comment]
I've tried to include it before your commit, no luck, it's still there. 😿

**Comment by tuexen** [comment]
OK, then I think the problem is in `user_atomic.h`:
```
#if defined(__Userspace_os_Windows)
#if defined(__MINGW32__)
static void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
#else
static inline void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
```
Can you figure out which prototype is needed on you platform and whether `__MINGW32__` is defined or not? Maybe it should be:
```
#if defined(__Userspace_os_Windows)
#if defined(__MINGW32__)
static void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#else
static inline void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
#else
static inline void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
```
or
 ```
#if defined(__Userspace_os_Windows)
#if defined(__MINGW32__) || defined(__MINGW64__)
static void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#else
static inline void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
#else
static inline void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
```

**Comment by nxrighthere** [comment]
I believe it should be:
```c
#if defined(__Userspace_os_Windows)
static void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#else
static inline void atomic_init(void) {} /* empty when we are not using atomic_mtx */
#endif
```

By the way, MinGW x64 is compatible with `__MINGW32__`.

There's are more errors appears along the road:
```E:\C\SCTP\usrsctplib\netinet\sctp_usrreq.c: In function 'sctp_finish':
E:\C\SCTP\usrsctplib\netinet\sctp_usrreq.c:201:39: error: comparison of integer expressions of different signedness: 'SOCKET' {aka 'long long unsigned int'} and 'int' [-Werror=sign-compare]
  if (SCTP_BASE_VAR(userspace_rawsctp) != -1) {
                                       ^~
E:\C\SCTP\usrsctplib\netinet\sctp_usrreq.c:209:39: error: comparison of integer expressions of different signedness: 'SOCKET' {aka 'long long unsigned int'} and 'int' [-Werror=sign-compare]
  if (SCTP_BASE_VAR(userspace_udpsctp) != -1) {
                                       ^~
E:\C\SCTP\usrsctplib\netinet\sctp_usrreq.c:219:40: error: comparison of integer expressions of different signedness: 'SOCKET' {aka 'long long unsigned int'} and 'int' [-Werror=sign-compare]
  if (SCTP_BASE_VAR(userspace_rawsctp6) != -1) {
                                        ^~
E:\C\SCTP\usrsctplib\netinet\sctp_usrreq.c:227:40: error: comparison of integer expressions of different signedness: 'SOCKET' {aka 'long long unsigned int'} and 'int' [-Werror=sign-compare]
  if (SCTP_BASE_VAR(userspace_udpsctp6) != -1) {
``

I think it's better to setup CI as Lennart suggested. 😸

**Comment by weinrank** [comment]
@nxrighthere Please test PR #329 

It seems that MinGW sets Windows 2003 Server as the oldest supported windows version.
Support for `inet_pton` and `inet_ntop` was introduced for Windows Vista and newer.
Since even Windows Vista is out of support, I've set the minimum supported (API) version to Windows 7. 

In addition I've fixed the compile issues you mentioned above.

**Comment by lgrahl** [comment]
At least the Meson MinGW build is fixed with #329.

**Comment by nxrighthere** [comment]
@weinrank Works like a charm, thanks!

> It seems that MinGW sets Windows 2003 Server as the oldest supported windows version.
Support for inet_pton and inet_ntop was introduced for Windows Vista and newer.
Since even Windows Vista is out of support, I've set the minimum supported (API) version to Windows 7.

Interesting, I didn't know about this. 🤔
